overlord/devicestate/users.go

```diff
@@ -0,0 +1,359 @@
+// user details are fetched from existing system user assertions.
+// If no email is passed, all known users will be created based on valid system user assertions.
+// If email is passed, only corresponding system user assertion is used for user creation.
+func CreateKnownUsers(st *state.State, sudoer bool, email string) ([]CreatedUser, error) {
+	model, err := findModel(st)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create user: cannot get model assertion: %v", err)
+	}
+
+	serial, err := findSerial(st, nil)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, fmt.Errorf("cannot create user: cannot get serial: %v", err)
+	}
+
+	db := assertstate.DB(st)
+	u := &createUserOpts{
+		assertDb: db,
+		modelAs:  model,
+		serialAs: serial,
+		isSudoer: sudoer,
+	}
+
+	if email == "" {
+		return u.createAllKnownSystemUsers(st)
+	}
+
+	username, opts, err := getUserDetailsFromAssertion(u.assertDb, u.modelAs, u.serialAs, email)
+	if err != nil {
+		return nil, &UserError{Err: err}
+	}
+
+	opts.Sudoer = sudoer
+	createdUser, err := u.createUser(st, opts, username, email)
+	if err != nil {
+		return nil, err
+	}
+	return []CreatedUser{createdUser}, nil
+}
+
+// RemoveUser removes linux user account of passed username.
+func RemoveUser(st *state.State, username string) (*auth.UserState, error) {
+	// TODO: allow to remove user entries by email as well
+
+	// catch silly errors
+	if username == "" {
+		return nil, &UserError{Err: fmt.Errorf("need a username to remove")}
+	}
+
+	// check the user is known to snapd
+	_, err := auth.UserByUsername(st, username)
+	if err == auth.ErrInvalidUser {
+		return nil, &UserError{Err: fmt.Errorf("user %q is not known", username)}
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// first remove the system user
+	if err := osutilDelUser(username, &osutil.DelUserOptions{ExtraUsers: !release.OnClassic}); err != nil {
+		return nil, err
+	}
+
+	// then the UserState
+	u, err := auth.RemoveUserByUsername(st, username)
+	// ErrInvalidUser means "not found" in this case
+	if err != nil && err != auth.ErrInvalidUser {
+		return nil, err
+	}
+	return u, nil
+}
+
+func getUserDetailsFromStore(theStore snapstate.StoreService, email string) (string, *osutil.AddUserOptions, error) {
+	v, err := theStore.UserInfo(email)
+	if err != nil {
+		return "", nil, fmt.Errorf("cannot create user %q: %s", email, err)
+	}
+	if len(v.SSHKeys) == 0 {
+		return "", nil, fmt.Errorf("cannot create user for %q: no ssh keys found", email)
+	}
+
+	// Amend information where the key came from to ensure it can
+	// be update/replaced later
+	for i, k := range v.SSHKeys {
+		v.SSHKeys[i] = fmt.Sprintf(`%s # snapd {"origin":"store","email":%q}`, k, email)
+	}
+
+	gecos := fmt.Sprintf("%s,%s", email, v.OpenIDIdentifier)
+	opts := &osutil.AddUserOptions{
+		SSHKeys: v.SSHKeys,
+		Gecos:   gecos,
+	}
+	return v.Username, opts, nil
+}
+
+// createUserOpts is a helper to handle user operations
+type createUserOpts struct {
```

Comment:
Despite being better as an options struct rather than a short-lived manager, it still seems odd to have this much logic as a method of the options instead of just passing an options struct into a function as we usually do. For context: https://github.com/snapcore/snapd/pull/11796#discussion_r943598250

Reply:
Good point!